sys/net/gnrc/transport_layer/tcp/gnrc_tcp.c

```diff
@@ -0,0 +1,630 @@
+    if (tcb->state != CLOSED) {
+        mutex_unlock(&(tcb->function_lock));
+        return -EISCONN;
+    }
+
+    /* Setup connection (common parts) */
+    msg_init_queue(tcb->msg_queue, GNRC_TCP_MSG_QUEUE_SIZE);
+    tcb->owner = thread_getpid();
+
+    /* Setup passive connection */
+    if (passive){
+        /* Set Status Flags */
+        tcb->status |= GNRC_TCP_STATUS_PASSIVE;
+        if (local_addr == NULL) {
+            tcb->status |= GNRC_TCP_STATUS_ALLOW_ANY_ADDR;
+        }
+        /* If local address is specified: Copy it into tcb: only connections to this addr are ok */
+        else {
+            switch (tcb->address_family) {
+#ifdef MODULE_GNRC_IPV6
+                case AF_INET6:
+                    memcpy(tcb->local_addr, local_addr, sizeof(ipv6_addr_t));
+                    break;
+#endif
+            }
+        }
+        /* Assign Port to listen on, to tcb */
+        *(tcb->local_port) = local_port;
+    }
+    /* Setup active connection */
+    else{
+        /* Copy Target Address and Port into tcb structure */
+        switch (tcb->address_family) {
+ #ifdef MODULE_GNRC_IPV6
+            case AF_INET6:
+                memcpy(tcb->peer_addr, target_addr, sizeof(ipv6_addr_t));
+                break;
+ #endif
+        }
+        /* Copy Port Information, verfication happens in fsm */
+        *(tcb->local_port) = local_port;
+        *(tcb->peer_port) = target_port;
+
+        /* Setup Timeout: If connection could not be established before */
+        /* the timer expired, the connection attempt failed */
+        connection_timeout_msg.type = MSG_TYPE_CONNECTION_TIMEOUT;
+        xtimer_set_msg(&connection_timeout_timer, GNRC_TCP_CONNECTION_TIMEOUT_DURATION,
+                       &connection_timeout_msg, tcb->owner);
+    }
+
+    /* Call FSM with Event: CALL_OPEN */
+    ret = _fsm(tcb, CALL_OPEN, NULL, NULL, 0);
+    if (ret == -ENOMEM) {
+        DEBUG("gnrc_tcp.c : gnrc_tcp_connect() : Out of receive buffers.\n");
+    } else if(ret == -EADDRINUSE) {
+        DEBUG("gnrc_tcp.c : gnrc_tcp_connect() : local_port is already in use.\n");
+    }
+
+    /* Wait until a connection was established or closed */
+    while (ret >= 0 && tcb->state != CLOSED
+           && tcb->state != ESTABLISHED && tcb->state != CLOSE_WAIT) {
+        msg_receive(&msg);
+        switch (msg.type) {
+            case MSG_TYPE_CONNECTION_TIMEOUT:
+                DEBUG("gnrc_tcp.c : _gnrc_tcp_open() : CONNECTION_TIMEOUT\n");
+                _fsm(tcb, TIMEOUT_CONNECTION, NULL, NULL, 0);
+                ret = -ETIMEDOUT;
+                break;
+
+            case MSG_TYPE_NOTIFY_USER:
+                DEBUG("gnrc_tcp.c : _gnrc_tcp_open() : MSG_TYPE_NOTIFY_USER\n");
+                break;
+
+            default:
+                DEBUG("gnrc_tcp.c : _gnrc_tcp_open() : other message type\n");
+        }
+    }
+
+    /* Cleanup */
+    if (ret != -ETIMEDOUT) {
+      xtimer_remove(&connection_timeout_timer);
```

Comment:
The return value is -ETIMEDOUT in case the timer fired before. In that case the timer does not need to stop. So I see no problem there.